[Enhancement] Introduce null_if_not_exist parameter for dictionary_get

srlch opened the pull request:
In current implementation, `dictionary_get` will throw an exception if the given key does not exist in dictionary cache. This pr introduce a new parameter called `null_if_not_exist` for `dictionary_get` to control the behavior when the key is not exist:

If `null_if_not_exist == false`(default value): throw an exception
If `null_if_not_exist == true`: return NULL for the unexisted key

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

## Documentation PRs only:

If you are submitting a PR that adds or changes English documentation and have not
included Chinese documentation, then you can check the box to request GPT to translate the
English doc to Chinese. Please ensure to uncheck the **Do not translate** box if translation is needed.
The workflow will generate a new PR with the Chinese translation after this PR is merged.

- [ ] Yes, translate English markdown files with GPT
- [x] Do not translate

trueeyu commented:
Can we reuse this SQLMode: `ALLOW_THROW_EXCEPTION` to solve this problem?

trueeyu commented on be/src/exprs/dictionary_get_expr.cpp:
```diff
@@ -24,7 +24,9 @@
 DictionaryGetExpr::DictionaryGetExpr(const TExprNode& node)
-        : Expr(node), _dictionary_get_expr(node.dictionary_get_expr) {}
+        : Expr(node), _dictionary_get_expr(node.dictionary_get_expr) {
+    _type = TypeDescriptor(TYPE_STRUCT);
```
Why not get type from TExprNode?

srlch replied:
Updated

trueeyu commented on be/src/exprs/dictionary_get_expr.cpp:
```diff
@@ -58,18 +60,24 @@ StatusOr<ColumnPtr> DictionaryGetExpr::evaluate_checked(ExprContext* context, Ch
     }
     value_chunk->reserve(size);
 
+    ColumnPtr null_column = UInt8Column::create(size, 0);
```
Why reserve value_chunk, but resize null_column?

srlch replied:
Because probe_given_dictionary_cache maybe not change null_column if null_if_not_exist = false.

trueeyu commented on be/src/storage/dictionary_cache_manager.h:
```diff
@@ -203,62 +207,94 @@ class DictionaryCacheImpl final : public DictionaryCache {
+                    null_if_not_exist ? _append_nullable(dest, null_column) : _append_value(dest, iter->second);
                     if constexpr (std::is_same_v<ValueCppType, Slice>) {
-                        value_encode_flags[i] = *(reinterpret_cast<uint8_t*>(iter->second.data) - 1);
+                        value_encode_flags[i] = null_if_not_exist
```
what does it mean here

srlch replied:
Flag value used for decoding

wyb commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -1985,12 +1985,17 @@ public Void visitDictionaryGetExpr(DictionaryGetExpr node, Scope context) {
             int paramDictionaryKeysSize = params.size() - 1;
-            if (paramDictionaryKeysSize != dictionaryKeysSize) {
+            if (!(paramDictionaryKeysSize == dictionaryKeysSize || paramDictionaryKeysSize == dictionaryKeysSize + 1)) {
                 throw new SemanticException("dictionary: " + dictionaryName + " has keys size: " +
```
should improve the error message, better to add expected param size

srlch replied:
Updated

srlch committed:
[Enhancement] Introduce null_if_not_exist parameter for dictionary_get

In current implementation, dictionary_get will throw an exception if the
given key does not exist in dictionary cache. This pr introduce a new
parameter called null_if_not_exist for dictionary_get to control the
behavior when the key is not exist:

If null_if_not_exist == false(default value): throw an exception
If null_if_not_exist == true: return NULL for the unexisted key

Signed-off-by: srlch <linzichao@starrocks.com>
